docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+Currently, the `SemanticModel` APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to be bound.
+To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves.
+Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis.
+This includes features like intellisense.
+3. Users should be able to get at both declared nullability for variables, as well as seeing the inferred nullability at each expression from flow analysis.
+
+## General Concepts
+
+Nullability information is exposed publicly as an enum, `Nullability`.
+There are 5 possible values:
+
+* `NotApplicable` - This is used when nullability information is requested about something that does not have a nullability, such as a statement, directive, or other similar constructs.
+* `NotAnalyzed` - This is used to fulfill priority #2.
+If the user requests information while asking for nullability to not be calculated, we'll use this to stand in for nullability.
+* `Unknown` - This is used for legacy C# code that does not know about nullability.
+This can appear in modern C# code that calls unannotated C#, or in certain generic variance scenarios where a common nullability cannot be inferred from the supplied arguments.
+For use in this document unknown is represented with the `~` character, but this character does not appear in source.
+* `NotNull` - This is used when we know the nullability is not null.
+* `MayBeNull` - This is used when we know the nullability may be null.
+
+Nullability information will be retrieved from the `SemanticModel`, much like type information.
+`GetSemanticModel` will return a `SemanticModel` that is aware of nullability information, and will force binding and calculation of nullability information when queried.
+We will offer a method off of `SemanticModel`, `GetSingleStatementSemanticModel()`, which will return a `SemanticModel` that does not calculate nullability information, for particularly performance-sensitive scenarios.
+This `SemanticModel` will share a cache with its parent, much like speculative `SemanticModel`s, and all nullabilities accessed on symbols retrieved from this `SemanticModel` will be `NotAnalyzed`.
+
+Generally speaking, when you want to determine the nullability of an `ITypeSymbol`, you need to check the containing context.
+The declaring symbol (e.g. `ILocalSymbol`, `IFieldSymbol`, etc.) will have the declared nullability of the type.
+`GetTypeInfo` will give you both the `ITypeSymbol` and the `Nullability` of an expression.
+Nested nullabilities for type parameters are contained on the containing `ITypeSymbol`.
```

Comment:
this mix seems problematic.  it means there's now two sets of data that has to be used as a 'source of truth' for nullability. We use symbols in some cases, but not all.  

Given that nullability information is stored on a symbol, how do you reconcile this with:

> These include reference equality for `ITypeSymbol`s that aren't generic, such as `string`.

While i get that this is about non-generics, how do you do equality properly *for* generics.  for example, say i have:

`IList<string?>` and `IList<string>`.  Even if do `.Equals`, will that return `true`?  If not, that seems like you will be breaking existing code that expects those to be the same.  

However, if you're ok with those symbols *not* being hte same, it's unclear to me why it's then a problem for something like `string?` and `string` to not be reference equals.